### PR TITLE
fix(adapters): complete dev sessions only on Stop or window death

### LINKED ISSUE #48: [BUG] Dev session reported completed from the idle-poll shortcut while the agent window is still alive
### Description

`GenericAdapter.wait_for_completion` (`src/automator/adapters/generic.py`) can report a dev session `completed` from the **idle-polling shortcut** while the native agent window is **still alive**. Once an idle-grace window is armed, each silent tick does a non-blocking result-artifact check and returns `completed` the moment the artifact exists (`generic.py:238-245`, and again at grace expiry, `generic.py:260-267`) — neither path confirms the window has actually exited (`alive` is computed at `generic.py:221` but only gates the *crash* branch). The stall fallback shares the hole: `_final` (`generic.py:361-362`) upgrades to `completed` on artifact presence alone.

Three aggravating facts (verified on current `main`):

1. **The premature `completed` kills the live agent.** `AgentAdapter.run` (`src/automator/adapters/base.py:91-96`) runs `finally: self.kill(handle)` — the moment the wait returns, the window is killed. So this is not just a race with post-session reads: the agent is **terminated mid-turn**, before its remaining commits/cleanup/transcript flush.
2. **The artifact check runs before the pane-activity re-arm** (`generic.py:238` vs `generic.py:251`), so even a session that is visibly streaming output — demonstrably working — is declared complete and killed on the next idle tick.
3. **Re-driven specs widen the race to most of the session.** `devcontract.find_result_artifact` (`devcontract.py:193-211`) qualifies any `spec-*.md` carrying an `## Auto Run Result` heading with mtime ≥ launch. Escalation re-arm (`reset_spec_status`) rewrites only the frontmatter `status:` and **leaves the stale `## Auto Run Result` section in the file**, so on a re-drive the spec qualifies as "terminal" from the agent's first mid-session save — not just in the final minutes.

The engine then enters post-session handling (spec-frontmatter reconcile, sprint sync, `verify_dev`, usage tally) against files the killed-or-still-writing agent left half-finished; a mis-read there feeds the **destructive auto-rollback path**.

Surfaced while prototyping the native-Windows multiplexer backend (slower ConPTY timing widens the race), but the bug is platform-neutral — the same interleaving exists on POSIX.

### Steps to reproduce

Timing race, so conceptual repro:

1. Start a dev session; let it end a turn without a terminal artifact (awaiting a long test, or an early turn-end) → the result-less Stop arms the idle-grace window (`generic.py:322`).
2. Let the stall nudge re-invoke it (`generic.py:278`); the agent resumes working in a new turn while `stall_deadline` stays armed.
3. Mid-turn the agent appends `## Auto Run Result` to the spec (a tool call, typically before final prose/cleanup) — or, on a re-driven escalated story, simply saves any edit to a spec that still carries the stale section.
4. The next idle tick finds the artifact → `wait_for_completion` returns `completed`, `run()`'s `finally` kills the still-working window, and the engine post-processes half-written state.

### Expected behavior

Stop events and window-death are the **only authoritative** completion signals. A terminal artifact observed while the window is still alive is advisory: note it, keep waiting, and complete only once the Stop lands or the window dies. The stall fallback in `_final` should likewise accept a result only when the window is no longer alive.

### Actual behavior

The idle-tick and grace-expiry paths return `completed` on artifact presence alone; the session window — possibly mid-turn and streaming — is then killed by `run()`'s `finally`, and the engine post-processes (and can roll back) a half-written attempt.

### Proposed resolution

- Gate the idle/grace terminal-artifact shortcuts on `not self._window_alive(handle)`; while the window is alive they only re-arm/annotate. (The normal Stop-path completion — artifact + Stop with the window alive — is untouched; artifacts are written by tool calls *inside* turns, so a legitimate artifact cannot first appear after the final Stop beyond the existing 15 s result grace.)
- `_final` gains an `accept_result` flag set to "window not alive" for the stall fallback, so a stall verdict can't silently upgrade to `completed` under a live window.
- Consider stripping the stale `## Auto Run Result` section when a spec is re-armed for re-drive, so a re-driven spec can't qualify as terminal from its first save.
- Keep Stop + window-death as the authoritative paths; add focused tests: an idle-tick result under a live window must not complete the session, and a grace-expiry result under a live window must not complete it.

### Which area is this for?

CLI adapters & profiles

### bmad-auto Version

0.7.12

### Which coding CLI are you using?

Claude (claude)

### Operating System

Other (platform-neutral; observed timing widens on Windows/ConPTY, same race on Linux/macOS)

### Confirm

- [x] I've searched for existing issues
- [x] I'm using the latest version

### PR DESCRIPTION
**Pull request by dracic**
## Summary

Fixes #48

`GenericAdapter.wait_for_completion` could return `completed` from the idle-tick and grace-expiry shortcuts on terminal-artifact presence alone, while the agent window was still alive — `run()`'s `finally: kill()` then terminated the agent mid-turn and the engine post-processed half-written state (feeding the destructive auto-rollback path).

**Changes:**
- **Stop events and window death are now the only authoritative completion signals.** Both artifact shortcuts in the idle branch are removed outright — the window is provably alive at those points (a dead window returns `crashed` earlier in the same tick), so the checks could never legitimately complete.
- `_final` gains a keyword-only `accept_result` flag; the grace-expiry stall verdict passes `False` so an artifact under a live window cannot silently upgrade `stalled` to `completed`. Window-death paths (`crashed`, `SessionEnd`) keep the trusting default — death is authoritative.
- **Re-drive amplifier closed:** `strip_auto_run_result` (new, in `devcontract`) removes the stale `## Auto Run Result` section when the escalation re-arm re-opens a spec, so a re-driven spec no longer qualifies as a terminal artifact from the agent's first mid-session save. The helper is fence-aware (incl. fences indented up to 3 spaces per CommonMark) since deletion is destructive; the read-only heading scans are intentionally untouched (follow-up: #52).
- The diagnostic trace scaffolding from the original prototype branch was deliberately **not** ported — it hunted an engine-death window already fixed on main.

**Behavioral notes:**
- Stop-path completion (artifact + Stop, window alive) is untouched; stall-nudge and pane-activity re-arm logic identical.
- One existing test amended to the new invariant: a completion marker landing after the nudge cap now completes via its turn-end Stop, not via the bare grace-expiry artifact.

## Tests

- New: idle-tick artifact under a live window does not complete (completes on the subsequent Stop); grace-expiry artifact under a live window stalls with `result_json=None`; window death with artifact still completes; strip unit tests incl. fenced/indented-fence edge cases; engine re-arm strips the stale section with the intent body untouched.
- Full suite green on Windows (1324 passed) and Linux/WSL (1335 passed); only the two pre-existing `test_module_skills_sync` drift failures, unrelated.

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevents stale terminal “Auto Run Result” sections from being treated as fresh completion when a spec is reopened or re-armed.
  * Improves fence-aware handling so “Auto Run Result” headings inside fenced code blocks are ignored for both detection and removal.
  * Enhances terminal-session completion reliability: late `result.json` no longer overrides an in-progress/stalled verdict while the window is alive; grace expiry and window death are honored correctly.
* **Tests**
  * Expanded coverage for fenced-heading edge cases, repair/re-arm stale-section stripping, and terminal completion timing/grace/liveness scenarios.

### CODE REVIEW

**Comment by pbean** [comment]
Thanks for the excellent PR — the issue write-up (#48), the adversarial follow-up (#52), and the test coverage made validation straightforward. Everything checked out: the removed shortcuts sit strictly after the same-tick liveness return so the window is provably alive there, all four profiles emit canonical Stop events, the base adapter never arms the stall grace (so triage/non-dev sessions were never on the removed paths), and window-death still honors the artifact.

I pushed two maintainer commits onto the branch (maintainer edits are enabled):

- **69dc4ed** — `runs.rearm_escalation` (the bmad-loop-resolve re-arm) still left the stale `## Auto Run Result` section behind; the strip only hooked `engine._reset_spec_for_repair`, which fires on the in-loop feedback-repair path, not the escalation re-drive. A resolved re-drive could therefore still parse the prior attempt's stale outcome (typically `Status: blocked`) on its first save + Stop. The re-arm now strips alongside the status flip, covering both re-arm paths as the PR description intended.
- **e782282** — folds in the fix for #52 while we're here: `parse_auto_run_result` / `find_result_artifact` now share the same fence-aware heading/boundary helpers as `strip_auto_run_result`, so a fence-quoted heading never qualifies a spec as terminal and the detectors no longer diverge from the stripper.

Full suite + trunk green locally on both commits; CI matrix re-running on the new head.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by pbean** [comment]
augment review

File: src/bmad_loop/devcontract.py [hunk 1]
```diff
@@ -83,21 +83,47 @@ class AutoRunResult:
     detail: str  # the prose body after the heading, trimmed (human-readable)
 
 
+def _fenced(text: str, offset: int) -> bool:
+    """True when `offset` falls inside a ``` / ~~~ fenced code block. Counting
+    fence lines before the offset suffices: an odd count means a fence is open.
+    Fences may be indented up to three spaces (CommonMark; a tab would make an
+    indented code block instead, so tabs are excluded)."""
+    return sum(1 for _ in re.finditer(r"^ {0,3}(?:```|~~~)", text[:offset], re.MULTILINE)) % 2 == 1
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/bmad_loop/devcontract.py]: Good catch — fixed in fdcb4c0.

`_fenced` no longer relies on a bare line-parity count. It now tracks the open fence's **character and length** and closes only on a later line using the same character, at least as long as the opener, with no trailing info string (CommonMark close rules). So a shorter ``` ``` ``` run inside a ```` ```` ```` block, or a ``` ``` ``` line inside a `~~~` block, stays literal content — a fenced `## Auto Run Result` is no longer misread as a real terminal section. That matters most on the destructive `strip_auto_run_result` path you flagged (and symmetrically stops a real section from being hidden behind a stray inner fence).

Added `parse`/`find_result_artifact`/`strip` tests for the nested-longer-fence and mismatched-fence-char cases, plus a closed-then-real positive control so the tracker can't over-correct into a fence that never ends. All five pre-existing fence tests (simple paired 3-backtick fences) are unchanged and green.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by pbean** [comment]
@coderabbit review
augment review

File: src/bmad_loop/adapters/generic.py [hunk 2]
```diff
@@ -279,7 +264,12 @@ def wait_for_completion(self, handle: SessionHandle, spec: SessionSpec) -> Sessi
                         stall_deadline = time.monotonic() + self._stall_grace_s
                         last_activity = self._log_activity_key(handle.task_id)
                         continue
-                    return self._final(handle, spec, "stalled", session_id, transcript_path)
+                    # The window is alive here, so an artifact on disk cannot
+                    # upgrade the stall verdict to completed — it may be stale
+                    # or mid-write; only a Stop or window death vouches for it.
+                    return self._final(
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/bmad_loop/adapters/generic.py:284]: Good call — fixed in 93a9e50.

You're right that the top-of-tick probe and this stall return share a tick, and because the return exits the loop a hard death (no `SessionEnd`) landing in the gap would never be caught by a later top-of-tick probe — so `accept_result=False` could discard a just-flushed artifact that window death is supposed to honor.

The stall path now re-probes `_window_alive` once immediately before finalizing; a now-dead window flows through the crash path (`accept_result=True`), so the outcome only ever *upgrades* — stall→crash, or stall→completed when the artifact is present. It can't regress the artifact-absent case, since `decide_dev`/`decide_review_session` branch on `status != "completed"` only, treating `crashed` and `stalled` identically. A transport error on the re-probe is treated as "not proof of death" (mirroring the top-of-tick handling) and falls through to the stall, with `spec.timeout_s` bounding a persistent failure.

Added two tests: a window that dies between the two probes on the same tick now completes on its artifact (and asserts both probes fired), and a transport error on the re-probe still stalls without crashing.

Note this narrowly closes the *in-tick* death race you flagged. The broader case — a session that genuinely finished but lost its Stop and stays window-*alive* through nudge exhaustion — is tracked separately as #61 (a guarded post-kill artifact re-scan) and is out of scope here.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by pbean** [comment]
augment review
@coderabbit review

File: src/bmad_loop/devcontract.py [hunk 3]
```diff
@@ -83,21 +83,70 @@ class AutoRunResult:
     detail: str  # the prose body after the heading, trimmed (human-readable)
 
 
+# A fence line: up to three spaces of indent, then a maximal run of >= 3 backticks
+# or tildes (its char AND length both matter per CommonMark), then the rest of the
+# line — an info string on an opener; on a close, only whitespace is allowed.
+_FENCE_LINE_RE = re.compile(r"^ {0,3}(`{3,}|~{3,})([^\n]*)$", re.MULTILINE)
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/bmad_loop/devcontract.py:89]: Thanks — I dug into this one carefully. It's safe as written; I've added a regression test to lock it in rather than change the parser.

The fence regex alone doesn't show the whole picture — there's a countervailing anchor. For `strip_auto_run_result` to remove anything, `AUTO_RUN_HEADING_RE` (`^##\s+Auto Run Result`) has to match, and it — like the section-boundary scan in `_next_heading_start` (`^##\s`) — is anchored at **column 0 with no leading-space tolerance**. Paired with `_FENCE_LINE_RE`'s 0–3-space bound, that covers both directions:

- A `## Auto Run Result` **genuinely nested** inside a list-indented (4+ space) fence is itself co-indented ≥1 space, so it never matches the column-0 heading regex — there's nothing to strip.
- A **column-0** `## Auto Run Result` is a real CommonMark ATX heading in every case *except* when it sits inside a 0–3-space fence — exactly the range `_fenced` already recognizes and excludes.

I checked the real-vs-fenced classification against CommonMark on the concrete cases (nested-list fence at col 4, a top-level 4-space indented code block, a col-0 heading inside a col-2 fence); it agrees on all of them, including your 4+-space example (0 sections detected → no-op strip).

I also weighed actually widening `_FENCE_LINE_RE` to 4+ spaces and decided against it: per CommonMark a fence indented 4+ spaces isn't a fence — it's an indented code block. Recognizing it as one would risk a *real* column-0 `## Auto Run Result` following an indented code block (with a stray ``` inside) being treated as fenced and **not** stripped — a false-negative that leaks a stale terminal section past the re-arm, which is the failure this PR exists to prevent. The 0–3 bound is the spec-faithful choice.

Added `test_strip_auto_run_result_ignores_list_indented_fence` (commit b0b4f8c) encoding your exact scenario as a no-op strip, so the invariant is guarded: if the heading regex ever gains leading-space tolerance, that test fails loudly.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by pbean** [comment]
augment review

File: src/bmad_loop/devcontract.py [hunk 4]
```diff
@@ -263,3 +313,30 @@ def _repl(m: re.Match[str]) -> str:
...
+
+    Companion to `reset_spec_status` on the re-drive path: re-opening a spec by
+    flipping only its frontmatter would leave the stale terminal section behind,
+    and `find_result_artifact` keys on that heading — the re-driven session's
+    very first save of the spec would then qualify as a terminal result. Each
+    section spans its heading to the next same-level heading (the shared
+    `parse_auto_run_result` boundary) or end-of-file; headings quoted inside
+    fenced code blocks are ignored on both ends. Returns True when a section was
+    removed, False when none was present."""
+    text = spec_path.read_text(encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/bmad_loop/devcontract.py:340]: Valid on the re-arm path — fixed in d8b6917, though I scoped the guard a bit differently than a blanket I/O wrapper.

You're right that `strip_auto_run_result` read/writes unconditionally. Tracing the two call sites:

- **`runs.rearm_escalation`** — `verify.set_frontmatter_status` already no-ops on a missing spec (`if not path.is_file(): return False`), so the following `strip_auto_run_result` was genuinely the first line to raise. Your concern lands directly here.
- **`engine._reset_spec_for_repair`** — the *sibling* `reset_spec_status` has the identical unguarded read and fires one line earlier, so guarding only the strip would have left this path still crashing.

So I guarded **both** `reset_spec_status` and `strip_auto_run_result` with the same `if not spec_path.is_file(): return False` the rest of the module already uses (`verify.set_frontmatter_status` / `read_frontmatter` / `synthesize_result`) — one convention, both call sites covered uniformly.

I deliberately kept the scope to the **absent-file** case rather than wrapping the read/write in try/except-and-no-op. A present-but-unreadable spec, or a write that fails mid-sequence, still raises on purpose: both callers flip the frontmatter status *before* stripping, so silently swallowing a strip failure would leave the spec re-opened but still carrying its stale `## Auto Run Result` section — which is exactly the state that makes the re-driven session's first save parse as a terminal result, the bug this PR exists to prevent. Absent → clean no-op; unreadable/failed-write → loud, so the stale section can't silently survive the re-open.

Added `test_reset_spec_status_noop_when_spec_absent` and `test_strip_auto_run_result_noop_when_spec_absent`; full `test_devcontract.py` (53) plus the `engine`/`runs`/`resolve` call-site suites (129) green.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by pbean** [comment]
augment review

File: src/bmad_loop/devcontract.py [hunk 5]
```diff
@@ -83,21 +83,70 @@ class AutoRunResult:
...
+    agent's first save (#52)."""
+    return [m for m in AUTO_RUN_HEADING_RE.finditer(text) if not _fenced(text, m.start())]
+
+
+def _next_heading_start(text: str, offset: int) -> int:
+    """Offset of the first non-fenced same-level (`## `) heading at/after
+    `offset`, or end-of-text — the shared section boundary. Fenced `## ` lines
+    inside the section (quoted shell comments, log output) are content, not
+    boundaries."""
+    for nxt in re.finditer(r"^##\s", text, re.MULTILINE):
```
[the remark this answers is not in the document]
Comment (pbean) [reply src/bmad_loop/devcontract.py:131]: Good eye on the `\s` — you're right it matches a newline, so `^##\s` does catch a bare `##` line. I dug in and it's correct as written; I added a lock-in test (commit `142a452`) rather than tighten the regex, because requiring a space/tab delimiter would regress here.

The crux: **a bare `##` is a valid empty ATX heading in CommonMark**, so `_next_heading_start` treating it as a same-level boundary is faithful, not premature. Tracing the three places the concern could bite:

- **`parse_auto_run_result`** — a bare `##` only bounds `detail` early. `detail` is human-readable routing prose (the blocked→PAUSE decision + the human-facing message); it's never a gate. The one structural field, `Status:`, is parsed from the top of the section, above any such line — unaffected.
- **terminal detection** — `find_result_artifact` / `_section_headings` key on `AUTO_RUN_HEADING_RE` (`^##\s+Auto Run Result`), never a bare `##`. A region after a bare `##` can't qualify as a result artifact, so the first-save-reads-as-terminal bug this PR exists to prevent is untouched.
- **`strip_auto_run_result`** (the destructive path) — bounding at the bare `##` makes the strip stop *earlier* and delete *less*. That's the safe direction.

I also checked the suggested `^##[ \t]` and it regresses rather than helps: it stops recognizing empty `##` headings as boundaries on **both** LF and CRLF (verified — end-of-line matching misses `\r\n`). On the strip path that means running *past* a real empty heading and over-deleting — the wrong direction for a destructive removal. The only genuine imprecision in `\s` is that it also covers `\f`/`\v`, which aren't valid CommonMark heading delimiters and never appear in a spec.

So I kept `\s` and added `test_parse_stops_at_bare_empty_heading` + `test_strip_auto_run_result_stops_at_bare_empty_heading` locking the boundary (parse `Status:` unaffected; strip keeps the empty-heading region). If the delimiter is ever tightened, those fail loudly.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by pbean** [comment]
augment review
